client/my-sites/checkout/checkout-thank-you/transfer-pending/index.tsx

```diff
@@ -0,0 +1,130 @@
+/**
+ * External dependencies
+ */
+import page from 'page';
+import * as React from 'react';
+import { sprintf } from '@wordpress/i18n';
+import { useI18n } from '@wordpress/react-i18n';
+import { useDispatch, useSelector } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import getAtomicTransfer from 'calypso/state/selectors/get-atomic-transfer';
+import { errorNotice } from 'calypso/state/notices/actions';
+import { transferStates } from 'calypso/state/atomic-transfer/constants';
+import { hideMasterbar, showMasterbar } from 'calypso/state/ui/masterbar-visibility/actions';
+import { useInterval } from '../../../../lib/interval/use-interval';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+// Total time to perform "loading"
+const DURATION_IN_MS = 60000;
+
+// Props type
+interface Props {
+	siteId: number;
+	orderId: number;
+}
+
+/* eslint-disable wpcalypso/jsx-classname-namespace */
+const TransferPending: React.FunctionComponent< Props > = ( props ) => {
+	const { __ } = useI18n();
+	const { siteId, orderId } = props;
+	const dispatch = useDispatch();
+	const siteSlug = useSelector( ( state ) => getSiteSlug( state, siteId ) );
+	const transfer = useSelector( ( state ) => getAtomicTransfer( state, siteId ) );
+
+	const steps = React.useRef< string[] >( [
+		__( 'Setting up your site' ),
+		__( 'Upgrading infrastructure' ),
+		__( 'Preparing WooCommerce' ),
+	] );
+
+	const totalSteps = steps.current.length;
+
+	const [ currentStep, setCurrentStep ] = React.useState( 0 );
+
+	/**
+	 * Completion progress: 0 <= progress <= 1
+	 */
+	const progress = ( currentStep + 1 ) / totalSteps;
+	const isComplete = progress >= 1;
+
+	useInterval(
+		() => setCurrentStep( ( s ) => s + 1 ),
+		// Enable the interval when progress is incomplete
+		isComplete ? null : DURATION_IN_MS / totalSteps
+	);
+
+	// Force animated progress bar to start at 0
+	const [ hasStarted, setHasStarted ] = React.useState( false );
+	React.useEffect( () => {
+		const id = setTimeout( () => setHasStarted( true ), 750 );
+		return () => clearTimeout( id );
+	}, [] );
+
+	// Hide toolbar while component is mounted
+	React.useEffect( () => {
+		dispatch( hideMasterbar() );
+		return () => {
+			dispatch( showMasterbar() );
+		};
+	}, [ dispatch ] );
+
+	// Redirect based on transfer status
+	React.useEffect( () => {
+		const retryOnError = () => {
+			page( `/stats/${ siteSlug }` );
+
+			dispatch(
```

Comment:
Minor, and I know that you said this was just copied from the old implementation but I think it's possible that this hook might run multiple times before its redirects complete since the `useEffect` will be run on every render and it's possible that there will be more than one render before `page()` gets the SPA to redirect. This isn't necessarily a problem, but it could possibly cause multiple versions of this error message to be displayed. One solution might be to have a `const didRedirect = useRef(false);` and then put `if (didRedirect.current) { return; }` at the start of the hook and `didRedirect.current = true;` just before each call to `page()`.

Also, I think the user might not see this error message since they're being redirected? It might be a good idea to add the `displayOnNextPage` option.

Finally, _just in case_ `page` happens to be really fast sometimes (maybe that doesn't happen?) it might be a good idea to put the notice dispatch _before_ the call to `page()`.

Reply:
Thanks for explaining the issue and rewriting this hook for the better! 😂 I made a couple of [additional changes](https://github.com/Automattic/wp-calypso/pull/55073/commits/8b6fd0e0f6abb6c9f8277140e82e5b8061ca234d):
- I added the `isPersistent` option in addition to `displayOnNextPage` since there seems to be some other routing going on as the component mounts. Now the error notice persists on the stats page:
<img width="1330" alt="Screen Shot 2021-08-03 at 2 16 40 PM" src="https://user-images.githubusercontent.com/1689238/128067811-ab5ec32a-9afd-4253-978f-30a189d71f1a.png">
- I added `setHasStarted( false );` before the call to `page()` since when `page` is slow we're still seeing part of the progress bar. That's a little tricky to test though.